File: modules/wem-api/src/main/java/com/enonic/wem/api/module/ModuleKey.java
package com.enonic.wem.api.module;

import org.osgi.framework.Bundle;
import org.osgi.framework.Version;

import com.google.common.base.Preconditions;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.base.Strings.isNullOrEmpty;
import static org.apache.commons.lang.StringUtils.substringAfterLast;
import static org.apache.commons.lang.StringUtils.substringBeforeLast;

public final class ModuleKey
{

    private static final String SYSTEM_MODULE_NAME = "system";

    public final static ModuleKey SYSTEM = new ModuleKey( ModuleName.from( SYSTEM_MODULE_NAME ), new ModuleVersion( "0.0.0" ) );

    private static final String SEPARATOR = "-";

    private final ModuleName name;

    private final ModuleVersion version;

    private final String refString;

    private ModuleKey( final ModuleName name, final ModuleVersion version )
    {
        checkNotNull( name );
        checkNotNull( version );
        this.name = name;
        this.version = version;
        this.refString = name.toString() + SEPARATOR + version.toString();
    }

    public ModuleName getName()
    {
        return name;
    }

    public ModuleVersion getVersion()
    {
        return version;
    }

    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( o == null || !( o instanceof ModuleKey ) )
        {
            return false;
        }
        final ModuleKey that = (ModuleKey) o;
        return refString.equals( that.refString );
    }

    @Override
    public int hashCode()
    {
        return refString.hashCode();
    }

    @Override
    public String toString()
    {
        return refString;
    }

    public static ModuleKey from( final ModuleName name, final ModuleVersion version )
    {
        return new ModuleKey( name, version );
    }

    public static ModuleKey from( final String moduleKey )
    {
        final String name = substringBeforeLast( moduleKey, SEPARATOR );
        if ( SYSTEM_MODULE_NAME.equals( moduleKey ) )
        {
            return ModuleKey.SYSTEM;
        }

        final String version = substringAfterLast( moduleKey, SEPARATOR );
        Preconditions.checkArgument( !isNullOrEmpty( version ), "Missing version in module key [" + moduleKey + "]" );

        final ModuleName moduleName = ModuleName.from( name );
        final ModuleVersion moduleVersion = ModuleVersion.from( version );
        return new ModuleKey( moduleName, moduleVersion );
    }

    public static ModuleKey from( final Bundle bundle )
    {
        final String name = bundle.getSymbolicName();
        final Version bundleVersion = bundle.getVersion();
        final ModuleVersion version = ModuleVersion.from( bundleVersion.toString() );
        return ModuleKey.from( ModuleName.from( name ), version );
    }
}

File: modules/wem-webapp/src/main/webapp/admin/common/js/module/ModuleKey.ts
module api.module {

    export class ModuleKey implements api.Equitable {

        private static SEPARATOR: string = "-";

        private name: string;

        private version: string;

        private refString: string;

        public static fromString(str: string): ModuleKey {
            var sepIndex: number = str.lastIndexOf(ModuleKey.SEPARATOR);
            if (sepIndex == -1) {
                throw new Error("ModuleKey must contain separator '" + ModuleKey.SEPARATOR + "':" + str);
            }

            var name = str.substring(0, sepIndex);
            var version = str.substring(sepIndex + 1, str.length);

            return new ModuleKey(name, version);
        }

        constructor(moduleName: string, moduleVersion: string) {
            this.name = moduleName;
            this.version = moduleVersion;
            this.refString = this.name + ModuleKey.SEPARATOR + this.version;
        }

        getName(): string {
            return this.name;
        }

        getVersion(): string {
            return this.version;
        }

        toString(): string {
            return this.refString;
        }

        equals(o: api.Equitable): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, ModuleKey)) {
                return false;
            }

            var other = <ModuleKey>o;

            if (!api.ObjectHelper.stringEquals(this.refString, other.refString)) {
                return false;
            }

            return true;
        }

        static toStringArray(keys: ModuleKey[]): string[] {
            var stringArray: string[] = [];
            keys.forEach((key: ModuleKey) => {
                stringArray.push(key.toString());
            });
            return stringArray;
        }

    }
}